doc/prebuilt/man/taler-exchange-closer.1
.\" Man page generated from reStructuredText.
.
.TH "TALER-EXCHANGE-CLOSER" "1" "Mar 22, 2020" "0.6pre1" "GNU Taler"
.SH NAME
taler-exchange-closer \- close idle reserves
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBtaler\-exchange\-closer\fP
[\fB\-h\fP\ |\ \fB\-\-help\fP] [\fB\-t\fP\ |\ \fB\-\-test\fP] [\fB\-v\fP\ |\ \fB\-\-version\fP]
.SH DESCRIPTION
.sp
\fBtaler\-exchange\-closer\fP is a command line tool to run close
reserves that have been idle for too long, causing transfers
to the originating bank account to be scheduled.
.INDENT 0.0
.TP
\fB\-c\fP \fIFILENAME\fP | \fB–config=\fP‌\fIFILENAME\fP
Use the configuration and other resources for the exchange to operate
from \fIFILENAME\fP\&.
.TP
\fB\-h\fP | \fB\-\-help\fP
Print short help on options.
.TP
\fB\-t\fP | \fB\-\-test\fP
Run in test mode and exit when idle.
.TP
\fB\-v\fP | \fB\-\-version\fP
Print version information.
.UNINDENT
.SH SEE ALSO
.sp
taler\-exchange\-transfer(1), taler\-exchange\-httpd(1), taler.conf(5).
.SH BUGS
.sp
Report bugs by using \fI\%https://gnunet.org/bugs/\fP or by sending electronic
mail to <\fI\%taler@gnu.org\fP>.
.SH AUTHOR
GNU Taler contributors
.SH COPYRIGHT
2014, 2015, 2016 Florian Dold, Benedikt Muller, Sree Harsha Totakura, Christian Grothoff, Marcello Stanisci (GPLv3+ or GFDL 1.3+)
.\" Generated by docutils manpage writer.
.


ChangeLog
Sat 14 Nov 2020 05:47:30 PM CET
    Modify taler-exchange-transfer to continue even after a
    wire transfer failed due to the bank refusing it because
    the target account does not exist. Changed the database
    to track such failures in the respective table.
    Opens new issue #6647.  -CG

Tue 10 Nov 2020 01:03:22 PM CET
    Updates to error codes and HTTP status codes for improved
    consistency. Fixed spelling issues. Ensure main() returns
    0 when called with '-h' or '--help'.
    Releasing GNU Taler 0.8.1. -CG

Sat 03 Oct 2020 03:59:45 PM CEST
    Various minor fixes, code cleanup, updates to more recent
    GNUnet APIs, new error codes, and timetravel test support.
    Releasing GNU Taler 0.8.0. -CG

Sun 29 Mar 2020 08:53:46 PM CEST
    Changed protocol to be more RESTful. Expanded auditor tests.
    Completed transition to new wire gateway API for bank interaction.
    Releasing GNU Taler 0.7.0. -CG

Tue 24 Dec 2019 11:09:14 PM CET
    Releasing GNU Taler 0.6.0. -CG

Sat 17 Aug 2019 10:03:38 PM CEST
    Remove "currency" field from exchange database, as we only
    support one currency per exchange anyway. -CG

Wed 26 Jun 2019 03:31:52 PM CEST
    Adding link signatures to prevent exchange from tracking
    users using coins falsely believed to have been recovered via /link,
    bumping protocol version to 4. -CG

Sat 08 Jun 2019 07:54:33 PM CEST
    Change payto://sepa/ to payto://iban/ as per current spec. -CG

Fri 03 May 2019 05:36:10 PM CEST
    Add support for compressing bodies during HTTP uploads to
    exchange and auditor. -CG

Thu 02 May 2019 09:15:37 PM CEST
    Pass hash of denomination public key in operations from
    clients to exchange instead of the (larger) full public key.
    Breaks protocol compatibility, bumping protocol to v3. -CG

Mon Mar 11 03:24:07 CET 2019
	Completed implementation of #5536 (delete private keys once we
	no longer need them). -CG

Sat Mar  2 19:09:43 CET 2019
	Changing denomination key revocation file format and moving them
	to their own directory (preparations for #5536 resolution). -CG

Sat Feb 23 14:47:12 CET 2019
	Actually disable 100 CONTINUE. -CG

Sun Aug 19 15:15:48 CEST 2018
	Increase various limits and rework transaction scopes to
	improve scalability. -CG

Tue Apr  3 23:29:06 CEST 2018
	Releasing Taler v0.5.0

Thu Jan  4 11:55:41 CET 2018
	Fix issue #5234 (aggregator ignoring refunds).
	Misc. minor fixes to the auditor. -CG

Mon Jan  1 23:15:37 CET 2018
	Add TALER_EXCHANGE_refund2() API call to libtalerexchange. -CG

Thu Dec 14 15:32:50 CET 2017
	Replaced taler-exchange-reservemod tool with new taler-bank-transfer
	tool (#5195).  Removed /admin/add/incoming API, replaced by new
	taler-exchange-wirewatch tooling. (#5077).

Sat Dec  9 15:21:50 CET 2071
	Implement refresh protocol optimization, bumping protocol version
	to v2 (#5178). -CG

Thu Nov  2 17:39:40 CET 2017
	Limit amount values to 2^53 as we always wanted (#5167). -CG

Thu Nov  2 16:50:46 CET 2017
	Fixing serious bug in reserve balance calculations, as
	fees were not always properly included in the summation. -CG

Tue Oct 31 13:43:55 CET 2017
	Extended crypto API to provide planchet generation functions
	explicitly (to be used on withdraw, refresh and tipping).
	Renamed a few API calls to better reflect their functions.
	Changed a few function signatures to use new structs. -CG

Wed Oct 18 15:20:29 CEST 2017
	Releasing Taler v0.4.0. -CG

Tue Oct 17 14:32:46 CEST 2017
	Fix building of /keys response to include full auditor
	signature set instead of partial map. -CG

Mon Oct 16 12:10:35 CEST 2017
	Realize JSON-based report formats from auditor (#4963). -CG

Sun Sep 17 16:46:13 CEST 2017
	Implement /keys cherry picking (#4840). -CG

Wed Sep 13 14:06:44 CEST 2017
	Add argument to TALER_EXCHANGE_check_keys_current()
	to force download even if /keys are still considered
	current. -CG

Fri Jul 14 17:38:54 CEST 2017
	Fix #4751 by not allowing denomination deletion (GC)
	to cascade into reserve_out table (and tolerating such
	deletion failures). This way, denominations will be kept
	around until the last reserve that references them (or
	known_coins derived from them) is closed. Naturally, in
	most case we expect reserves to be closed much faster
	than DKs, so in practice this should rarely apply. -CG

Sun Jun 11 17:03:56 CEST 2017
	Finish implementation and testing of automated refunding
	of expired reserves (#4956). -CG

Tue Jun  6 13:53:34 CEST 2017
	Releasing taler-exchange 0.3.0. -CG

Mon Apr 17 01:29:07 CEST 2017
	Add support for HTTP body compression (#4982). -CG

Mon Mar 20 04:37:46 CET 2017
	Implemented first working version of taler-auditor. -CG

Sat Mar 18 03:44:38 CET 2017
	Add 'wire_out' table to exchange DB to track outgoing
	wire transfers. -CG

Fri Nov 18 18:53:30 CET 2016
	Improved error reporting for bogus wire specifications.
	Releasing taler-exchange 0.2.0. -CG

Thu Nov 17 16:37:22 CET 2016
	Added missing idempotency checks for /refresh/reveal. -CG

Thu Nov 17 11:37:56 CET 2016
	Fixed a few cases of missing database rollbacks, causing the
	exchange to be stuck. -CG

Mon Jun 20 17:40:49 CEST 2016
 	Releasing taler-exchange 0.0.2. -CG

Mon Jun 13 16:35:09 CEST 2016
	Splitting HTTP service to run on different listen
	sockets for /admin and /test-APIs and the normal public
	APIs (#4569). -CG

Thu Jun  9 19:17:02 CEST 2016
	Rename /deposit/wtid to /track/transaction and
	/wire/deposits to /track/transfer. -CG

Wed Jun  8 11:32:41 CEST 2016
	Export libtalerfakebank API so we can use it in testcases for the
	merchant as well. -CG

Tue Jun  7 15:13:46 CEST 2016
        Adding public key of the exchange that was used to sign replies
        to applicable callbacks of libtalerexchange. (This will eventually
        be needed by the merchant's backend.) -CG

Mon Jun  6 23:59:05 CEST 2016
	Minor code cleanup, fixing FTBFS when run with just 'make'
	or when header installed to a location not included.
	Releasing taler-exchange 0.0.1. -CG


src/exchange/taler-exchange-closer.c
/*
  This file is part of TALER
  Copyright (C) 2016-2020 Taler Systems SA

  TALER is free software; you can redistribute it and/or modify it under the
  terms of the GNU Affero General Public License as published by the Free Software
  Foundation; either version 3, or (at your option) any later version.

  TALER is distributed in the hope that it will be useful, but WITHOUT ANY
  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
  A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.

  You should have received a copy of the GNU Affero General Public License along with
  TALER; see the file COPYING.  If not, see <http://www.gnu.org/licenses/>
*/

/**
 * @file taler-exchange-closer.c
 * @brief Process that closes expired reserves
 * @author Christian Grothoff
 */
#include "platform.h"
#include <gnunet/gnunet_util_lib.h>
#include <jansson.h>
#include <pthread.h>
#include "taler_exchangedb_lib.h"
#include "taler_exchangedb_plugin.h"
#include "taler_json_lib.h"
#include "taler_bank_service.h"


/**
 * What is the smallest unit we support for wire transfers?
 * We will need to round down to a multiple of this amount.
 */
static struct TALER_Amount currency_round_unit;

/**
 * What is the base URL of this exchange?  Used in the
 * wire transfer subjects to that merchants and governments
 * can ask for the list of aggregated deposits.
 */
static char *exchange_base_url;

/**
 * The exchange's configuration.
 */
static const struct GNUNET_CONFIGURATION_Handle *cfg;

/**
 * Our database plugin.
 */
static struct TALER_EXCHANGEDB_Plugin *db_plugin;

/**
 * Next task to run, if any.
 */
static struct GNUNET_SCHEDULER_Task *task;

/**
 * How long should we sleep when idle before trying to find more work?
 */
static struct GNUNET_TIME_Relative aggregator_idle_sleep_interval;

/**
 * Value to return from main(). 0 on success, non-zero
 * on serious errors.
 */
static enum
{
  GR_SUCCESS = 0,
  GR_WIRE_ACCOUNT_NOT_CONFIGURED = 1,
  GR_WIRE_TRANSFER_FEES_NOT_CONFIGURED = 2,
  GR_FAILURE_TO_ROUND_AMOUNT = 3,
  GR_DATABASE_INSERT_HARD_FAIL = 4,
  GR_DATABASE_SELECT_HARD_FAIL = 5,
  GR_DATABASE_COMMIT_HARD_FAIL = 6,
  GR_DATABASE_SESSION_START_FAIL = 7,
  GR_DATABASE_TRANSACTION_BEGIN_FAIL = 8,
  GR_CONFIGURATION_INVALID = 9,
  GR_CMD_LINE_UTF8_ERROR = 10,
  GR_CMD_LINE_OPTIONS_WRONG = 11,
  GR_INVALID_PAYTO_ENCOUNTERED = 12,
} global_ret;

/**
 * #GNUNET_YES if we are in test mode and should exit when idle.
 */
static int test_mode;


/**
 * Main work function that finds and triggers transfers for reserves
 * closures.
 *
 * @param cls closure
 */
static void
run_reserve_closures (void *cls);


/**
 * We're being aborted with CTRL-C (or SIGTERM). Shut down.
 *
 * @param cls closure
 */
static void
shutdown_task (void *cls)
{
  (void) cls;
  GNUNET_log (GNUNET_ERROR_TYPE_INFO,
              "Running shutdown\n");
  if (NULL != task)
  {
    GNUNET_SCHEDULER_cancel (task);
    task = NULL;
  }
  TALER_EXCHANGEDB_plugin_unload (db_plugin);
  db_plugin = NULL;
  TALER_EXCHANGEDB_unload_accounts ();
  cfg = NULL;
}


/**
 * Parse the configuration for wirewatch.
 *
 * @return #GNUNET_OK on success
 */
static int
parse_wirewatch_config (void)
{
  if (GNUNET_OK !=
      GNUNET_CONFIGURATION_get_value_string (cfg,
                                             "exchange",
                                             "BASE_URL",
                                             &exchange_base_url))
  {
    GNUNET_log_config_missing (GNUNET_ERROR_TYPE_ERROR,
                               "exchange",
                               "BASE_URL");
    return GNUNET_SYSERR;
  }
  if (GNUNET_OK !=
      GNUNET_CONFIGURATION_get_value_time (cfg,
                                           "exchange",
                                           "AGGREGATOR_IDLE_SLEEP_INTERVAL",
                                           &aggregator_idle_sleep_interval))
  {
    GNUNET_log_config_missing (GNUNET_ERROR_TYPE_ERROR,
                               "exchange",
                               "AGGREGATOR_IDLE_SLEEP_INTERVAL");
    return GNUNET_SYSERR;
  }
  if ( (GNUNET_OK !=
        TALER_config_get_amount (cfg,
                                 "taler",
                                 "CURRENCY_ROUND_UNIT",
                                 &currency_round_unit)) ||
       ( (0 != currency_round_unit.fraction) &&
         (0 != currency_round_unit.value) ) )
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Need non-zero amount in section `TALER' under `CURRENCY_ROUND_UNIT'\n");
    return GNUNET_SYSERR;
  }

  if (NULL ==
      (db_plugin = TALER_EXCHANGEDB_plugin_load (cfg)))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Failed to initialize DB subsystem\n");
    return GNUNET_SYSERR;
  }
  if (GNUNET_OK !=
      TALER_EXCHANGEDB_load_accounts (cfg))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "No wire accounts configured for debit!\n");
    TALER_EXCHANGEDB_plugin_unload (db_plugin);
    db_plugin = NULL;
    return GNUNET_SYSERR;
  }
  return GNUNET_OK;
}


/**
 * Perform a database commit. If it fails, print a warning.
 *
 * @param session session to perform the commit for.
 * @return status of commit
 */
static enum GNUNET_DB_QueryStatus
commit_or_warn (struct TALER_EXCHANGEDB_Session *session)
{
  enum GNUNET_DB_QueryStatus qs;

  qs = db_plugin->commit (db_plugin->cls,
                          session);
  if (GNUNET_DB_STATUS_SUCCESS_NO_RESULTS == qs)
    return qs;
  GNUNET_log ((GNUNET_DB_STATUS_SOFT_ERROR == qs)
              ? GNUNET_ERROR_TYPE_INFO
              : GNUNET_ERROR_TYPE_ERROR,
              "Failed to commit database transaction!\n");
  return qs;
}


/**
 * Closure for #expired_reserve_cb().
 */
struct ExpiredReserveContext
{

  /**
   * Database session we are using.
   */
  struct TALER_EXCHANGEDB_Session *session;

};


/**
 * Function called with details about expired reserves.
 * We trigger the reserve closure by inserting the respective
 * closing record and prewire instructions into the respective
 * tables.
 *
 * @param cls a `struct ExpiredReserveContext *`
 * @param reserve_pub public key of the reserve
 * @param left amount left in the reserve
 * @param account_payto_uri information about the bank account that initially
 *        caused the reserve to be created
 * @param expiration_date when did the reserve expire
 * @return transaction status code
 */
static enum GNUNET_DB_QueryStatus
expired_reserve_cb (void *cls,
                    const struct TALER_ReservePublicKeyP *reserve_pub,
                    const struct TALER_Amount *left,
                    const char *account_payto_uri,
                    struct GNUNET_TIME_Absolute expiration_date)
{
  struct ExpiredReserveContext *erc = cls;
  struct TALER_EXCHANGEDB_Session *session = erc->session;
  struct GNUNET_TIME_Absolute now;
  struct TALER_WireTransferIdentifierRawP wtid;
  struct TALER_Amount amount_without_fee;
  struct TALER_Amount closing_fee;
  int ret;
  enum GNUNET_DB_QueryStatus qs;
  struct TALER_EXCHANGEDB_WireAccount *wa;

  /* NOTE: potential optimization: use custom SQL API to not
     fetch this: */
  GNUNET_log (GNUNET_ERROR_TYPE_INFO,
              "Processing reserve closure at %s\n",
              GNUNET_STRINGS_absolute_time_to_string (expiration_date));
  now = GNUNET_TIME_absolute_get ();
  (void) GNUNET_TIME_round_abs (&now);

  /* lookup account we should use */
  wa = TALER_EXCHANGEDB_find_account_by_payto_uri (account_payto_uri);
  if (NULL == wa)
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "No wire account configured to deal with target URI `%s'\n",
                account_payto_uri);
    global_ret = GR_WIRE_ACCOUNT_NOT_CONFIGURED;
    GNUNET_SCHEDULER_shutdown ();
    return GNUNET_DB_STATUS_HARD_ERROR;
  }

  /* lookup `closing_fee` from time of actual reserve expiration
     (we may be lagging behind!) */
  {
    struct TALER_Amount wire_fee;
    struct GNUNET_TIME_Absolute start_date;
    struct GNUNET_TIME_Absolute end_date;
    struct TALER_MasterSignatureP master_sig;
    enum GNUNET_DB_QueryStatus qs;

    qs = db_plugin->get_wire_fee (db_plugin->cls,
                                  session,
                                  wa->method,
                                  expiration_date,
                                  &start_date,
                                  &end_date,
                                  &wire_fee,
                                  &closing_fee,
                                  &master_sig);
    if (0 >= qs)
    {
      GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                  "Could not get wire fees for %s at %s. Aborting run.\n",
                  wa->method,
                  GNUNET_STRINGS_absolute_time_to_string (expiration_date));
      return GNUNET_DB_STATUS_HARD_ERROR;
    }
  }

  /* calculate transfer amount */
  ret = TALER_amount_subtract (&amount_without_fee,
                               left,
                               &closing_fee);
  if ( (GNUNET_SYSERR == ret) ||
       (GNUNET_NO == ret) )
  {
    /* Closing fee higher than or equal to remaining balance, close
       without wire transfer. */
    closing_fee = *left;
    GNUNET_assert (GNUNET_OK ==
                   TALER_amount_get_zero (left->currency,
                                          &amount_without_fee));
  }
  /* round down to enable transfer */
  if (GNUNET_SYSERR ==
      TALER_amount_round_down (&amount_without_fee,
                               &currency_round_unit))
  {
    GNUNET_break (0);
    global_ret = GR_FAILURE_TO_ROUND_AMOUNT;
    GNUNET_SCHEDULER_shutdown ();
    return GNUNET_DB_STATUS_HARD_ERROR;
  }
  if ( (0 == amount_without_fee.value) &&
       (0 == amount_without_fee.fraction) )
    ret = GNUNET_NO;

  /* NOTE: sizeof (*reserve_pub) == sizeof (wtid) right now, but to
     be future-compatible, we use the memset + min construction */
  memset (&wtid,
          0,
          sizeof (wtid));
  memcpy (&wtid,
          reserve_pub,
          GNUNET_MIN (sizeof (wtid),
                      sizeof (*reserve_pub)));
  if (GNUNET_SYSERR != ret)
    qs = db_plugin->insert_reserve_closed (db_plugin->cls,
                                           session,
                                           reserve_pub,
                                           now,
                                           account_payto_uri,
                                           &wtid,
                                           left,
                                           &closing_fee);
  else
    qs = GNUNET_DB_STATUS_HARD_ERROR;
  GNUNET_log (GNUNET_ERROR_TYPE_INFO,
              "Closing reserve %s over %s (%d, %d)\n",
              TALER_B2S (reserve_pub),
              TALER_amount2s (left),
              ret,
              qs);
  /* Check for hard failure */
  if ( (GNUNET_SYSERR == ret) ||
       (GNUNET_DB_STATUS_HARD_ERROR == qs) )
  {
    GNUNET_break (0);
    global_ret = GR_DATABASE_INSERT_HARD_FAIL;
    GNUNET_SCHEDULER_shutdown ();
    return GNUNET_DB_STATUS_HARD_ERROR;
  }
  if ( (GNUNET_OK != ret) ||
       (GNUNET_DB_STATUS_SUCCESS_ONE_RESULT != qs) )
  {
    /* Reserve balance was almost zero OR soft error */
    GNUNET_log (GNUNET_ERROR_TYPE_INFO,
                "Reserve was virtually empty, moving on\n");
    (void) commit_or_warn (session);
    return qs;
  }

  /* success, perform wire transfer */
  {
    void *buf;
    size_t buf_size;

    TALER_BANK_prepare_transfer (account_payto_uri,
                                 &amount_without_fee,
                                 exchange_base_url,
                                 &wtid,
                                 &buf,
                                 &buf_size);
    /* Commit our intention to execute the wire transfer! */
    qs = db_plugin->wire_prepare_data_insert (db_plugin->cls,
                                              session,
                                              wa->method,
                                              buf,
                                              buf_size);
    GNUNET_free (buf);
  }
  if (GNUNET_DB_STATUS_HARD_ERROR == qs)
  {
    GNUNET_break (0);
    global_ret = GR_DATABASE_INSERT_HARD_FAIL;
    GNUNET_SCHEDULER_shutdown ();
    return GNUNET_DB_STATUS_HARD_ERROR;
  }
  if (GNUNET_DB_STATUS_SOFT_ERROR == qs)
  {
    /* start again */
    return GNUNET_DB_STATUS_SUCCESS_NO_RESULTS;
  }
  return GNUNET_DB_STATUS_SUCCESS_ONE_RESULT;
}


/**
 * Main work function that finds and triggers transfers for reserves
 * closures.
 *
 * @param cls closure
 */
static void
run_reserve_closures (void *cls)
{
  struct TALER_EXCHANGEDB_Session *session;
  enum GNUNET_DB_QueryStatus qs;
  struct ExpiredReserveContext erc;
  struct GNUNET_TIME_Absolute now;

  (void) cls;
  task = NULL;
  if (NULL == (session = db_plugin->get_session (db_plugin->cls)))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Failed to obtain database session!\n");
    global_ret = GR_DATABASE_SESSION_START_FAIL;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }

  if (GNUNET_OK !=
      db_plugin->start (db_plugin->cls,
                        session,
                        "aggregator reserve closures"))
  {
    GNUNET_log (GNUNET_ERROR_TYPE_ERROR,
                "Failed to start database transaction!\n");
    global_ret = GR_DATABASE_TRANSACTION_BEGIN_FAIL;
    GNUNET_SCHEDULER_shutdown ();
    return;
  }
  erc.session = session;
  now = GNUNET_TIME_absolute_get ();
  (void) GNUNET_TIME_round_abs (&now);
  GNUNET_log (GNUNET_ERROR_TYPE_INFO,
              "Checking for reserves to close by date %s\n",
              GNUNET_STRINGS_absolute_time_to_string (now));
  qs = db_plugin->get_expired_reserves (db_plugin->cls,
                                        session,
                                        now,
                                        &expired_reserve_cb,
                                        &erc);
  GNUNET_assert (1 >= qs);
  switch (qs)
  {
  case GNUNET_DB_STATUS_HARD_ERROR:
    GNUNET_break (0);
    db_plugin->rollback (db_plugin->cls,
                         session);
    global_ret = GR_DATABASE_SELECT_HARD_FAIL;
    GNUNET_SCHEDULER_shutdown ();
    return;
  case GNUNET_DB_STATUS_SOFT_ERROR:
    db_plugin->rollback (db_plugin->cls,
                         session);
    GNUNET_assert (NULL == task);
    task = GNUNET_SCHEDULER_add_now (&run_reserve_closures,
                                     NULL);
    return;
  case GNUNET_DB_STATUS_SUCCESS_NO_RESULTS:
    GNUNET_log (GNUNET_ERROR_TYPE_INFO,
                "No more idle reserves to close, going to sleep.\n");
    db_plugin->rollback (db_plugin->cls,
                         session);
    GNUNET_assert (NULL == task);
    if (GNUNET_YES == test_mode)
    {
      GNUNET_SCHEDULER_shutdown ();
    }
    else
    {
      task = GNUNET_SCHEDULER_add_delayed (aggregator_idle_sleep_interval,
                                           &run_reserve_closures,
                                           NULL);
    }
    return;
  case GNUNET_DB_STATUS_SUCCESS_ONE_RESULT:
    (void) commit_or_warn (session);
    GNUNET_assert (NULL == task);
    task = GNUNET_SCHEDULER_add_now (&run_reserve_closures,
                                     NULL);
    return;
  }
}


/**
 * First task.  Parses the configuration and starts the
 * main loop of #run_reserve_closures(). Also schedules
 * the #shutdown_task() to clean up.
 *
 * @param cls closure, NULL
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param c configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *c)
{
  (void) cls;
  (void) args;
  (void) cfgfile;

  cfg = c;
  if (GNUNET_OK != parse_wirewatch_config ())
  {
    cfg = NULL;
    global_ret = GR_CONFIGURATION_INVALID;
    return;
  }
  GNUNET_assert (NULL == task);
  task = GNUNET_SCHEDULER_add_now (&run_reserve_closures,
                                   NULL);
  GNUNET_SCHEDULER_add_shutdown (&shutdown_task,
                                 cls);
}


/**
 * The main function of the taler-exchange-closer.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, non-zero on error
 */
int
main (int argc,
      char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_timetravel ('T',
                                     "timetravel"),
    GNUNET_GETOPT_option_flag ('t',
                               "test",
                               "run in test mode and exit when idle",
                               &test_mode),
    GNUNET_GETOPT_OPTION_END
  };
  enum GNUNET_GenericReturnValue ret;

  if (GNUNET_OK !=
      GNUNET_STRINGS_get_utf8_args (argc, argv,
                                    &argc, &argv))
    return GR_CMD_LINE_UTF8_ERROR;
  ret = GNUNET_PROGRAM_run (
    argc, argv,
    "taler-exchange-closer",
    gettext_noop ("background process that closes expired reserves"),
    options,
    &run, NULL);
  GNUNET_free_nz ((void *) argv);
  if (GNUNET_SYSERR == ret)
    return GR_CMD_LINE_OPTIONS_WRONG;
  if (GNUNET_NO == ret)
    return 0;
  return global_ret;
}


/* end of taler-exchange-closer.c */
